api/src/main/java/org/openmrs/module/appframework/aop/LoginLocationsAdvisor.java

```diff
@@ -0,0 +1,75 @@
+ * http://license.openmrs.org
+ *
+ * Software distributed under the License is distributed on an "AS IS"
+ * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
+ * License for the specific language governing rights and limitations
+ * under the License.
+ *
+ * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
+ */
+package org.openmrs.module.appframework.aop;
+
+import org.aopalliance.aop.Advice;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.openmrs.Location;
+import org.openmrs.LocationTag;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.appframework.AppFrameworkConstants;
+import org.openmrs.module.appframework.repository.AllLoginLocations;
+import org.openmrs.module.appframework.service.AppFrameworkService;
+import org.springframework.aop.Advisor;
+import org.springframework.aop.AfterReturningAdvice;
+import org.springframework.aop.support.StaticMethodMatcherPointcutAdvisor;
+import java.lang.reflect.Method;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.ArrayList;
+
+public class LoginLocationsAdvisor extends StaticMethodMatcherPointcutAdvisor implements Advisor {
+
+    private static final Log log = LogFactory.getLog(LoginLocationsAdvisor.class);
+    Set<String> loginLocationsUpdateTriggers = new HashSet<String>();
+
+    public void LocationSearchAdviser() {
+        loginLocationsUpdateTriggers.add("saveLocation");
+        loginLocationsUpdateTriggers.add("retireLocation");
+        loginLocationsUpdateTriggers.add("unretireLocation");
+        
+    }
+
+    @Override
+    public boolean matches(Method method, Class targetClass) {
+        if(loginLocationsUpdateTriggers.contains(method.getName())) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public Advice getAdvice() {
+        return new LoginLocationsAdvice();
+    }
+
+    private class LoginLocationsAdvice implements AfterReturningAdvice {
+        public void afterReturning(Object returnValue, Method method, Object[] args, Object target) throws Throwable {
+            Set<LocationTag> tags = ((Location) returnValue).getTags();
+            for (Iterator<LocationTag> tagsIterator = tags.iterator(); tagsIterator.hasNext(); ) {
+                if (AppFrameworkConstants.LOCATION_TAG_SUPPORTS_LOGIN_UUID.equals(tagsIterator.next().getUuid())) {
+                    // The login locations are first set in AppFrameworkActivator
+                    AllLoginLocations allLoginLocations = Context.getRegisteredComponents(AllLoginLocations.class).get(0);
+                    allLoginLocations.clear();
+                    allLoginLocations.add(((AppFrameworkService) Context.getRegisteredComponent("appFrameworkService", AppFrameworkService.class)).getLoginLocations());
```

Comment:
I would try to just add/edit/remove the one added/edited/removed location.

Can you create a Spring test for showing how this works?